hs/src/Reach/Proto.hs

```diff
@@ -0,0 +1,189 @@
+{-# LANGUAGE DataKinds     #-}
+{-# LANGUAGE TypeOperators #-}
+module Reach.Proto
+  ( FDescOut'(..)
+  , FDescOut(..)
+  , Req(..)
+  , Proto
+  , cmd
+  , say
+  , listen
+  , runStubServer
+  , stubPort
+  ) where
+
+import Control.Concurrent
+import Control.Monad.IO.Class
+import Data.Aeson
+import Data.Proxy
+import Data.Text
+import Data.Time.Clock
+import Data.Typeable
+import GHC.Generics
+import Servant
+import Servant.Client
+import Servant.Types.SourceT
+import Text.Parsec
+import Text.Parsec.Language
+import Text.Parsec.Token
+import qualified Data.ByteString.Lazy as BL
+import qualified Data.ByteString.Lazy.UTF8 as BL
+import qualified Data.Map.Strict as M
+import qualified Network.HTTP.Media as H
+import qualified Network.Wai.Handler.Warp as Warp
+import qualified Servant.Client.Streaming as S
+
+-- https://docs.servant.dev/en/latest/cookbook/basic-streaming/Streaming.html
+-- https://docs.servant.dev/en/stable/tutorial/Server.html#streaming-endpoints
+-- https://docs.servant.dev/en/stable/tutorial/Client.html#querying-streaming-apis
+-- https://hackage.haskell.org/package/servant-0.19/docs/Servant-Types-SourceT.html#t:StepT
+--
+-- TODO much of the following ought to be renamed
+-- TODO "AppT" over `say`, `listen`, and `cmd` request helpers
+-- TODO easy thread-safety
+-- TODO prolong client timeouts
+
+stubPort :: Int
+stubPort = 8123
+
+data EventStream deriving Typeable
+
+data FDescOut'
+  = Stdout'
+  | Stderr'
+  | Stdboth'
+  deriving Eq
+
+instance FromHttpApiData FDescOut' where
+  parseUrlPiece = \case
+    "1" -> Right Stdout'
+    "2" -> Right Stderr'
+    "3" -> Right Stdboth'
+    a -> Left $ "Invalid fd: " <> a
+
+instance ToHttpApiData FDescOut' where
+  toUrlPiece = \case
+    Stdout'  -> "1"
+    Stderr'  -> "2"
+    Stdboth' -> "3"
+
+data FDescOut
+  = Stdout Integer BL.ByteString
+  | Stderr Integer BL.ByteString
+  | Keepalive
+
+type Project = M.Map FilePath Text -- path:hash
+
+data Req = Req
+  { req_args  :: [Text]
+  , req_files :: Maybe Project
+  , req_env   :: M.Map Text Text
+  } deriving (Eq, Generic, FromJSON, ToJSON)
+
+data Res = Res -- TODO
+  deriving (Eq, Generic, FromJSON, ToJSON)
+
+type Pid = Text
+
+instance Accept EventStream where
+  contentType _ = "text" H.// "event-stream"
+
+class ToSSE a where
+  toSSE :: a -> BL.ByteString
+
+toSSE' :: Show a => BL.ByteString -> a -> BL.ByteString -> BL.ByteString
+toSSE' e n m
+  = "id: " <> BL.fromString (show n) <> "\n"
+ <> "event: " <> e <> "\n"
+ <> "data: " <> m <> "\n\n"
+
+instance ToSSE FDescOut where
+  toSSE = \case
+    Stdout n m -> toSSE' "stdout" n m
+    Stderr n m -> toSSE' "stderr" n m
+    Keepalive  -> ": keepalive\n\n"
+
+instance ToSSE a => MimeRender EventStream a where
```

Comment:
We have `FlexibleInstances` on by default already? I don't feel strongly enough about it to object, but it makes me suspicious to see instances that are not defined on concrete types.

Reply:
https://github.com/reach-sh/reach-lang/blob/b9c95e61ab571309b4c1a3616a3d5e0c82e7a0b7/hs/package.mo.yaml#L109

I don't really have an opinion so if you'd like to see this changed I think we can.